segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/SegmentMetadataUpdateTransaction.java

```diff
@@ -450,9 +463,61 @@ private void preProcessAttributes(Collection<AttributeUpdate> attributeUpdates)
                 default:
                     throw new BadAttributeUpdateException(this.name, u, !hasValue, "Unexpected update type: " + updateType);
             }
+
+            // For updates by reference, we need to set the actual value.
+            if (byRef) {
+                u.setValue(newValue);
```

Comment:
For updates by reference, the value is not set. We are given a function to evaluate. At this point the function has already been evaluated and we need to set the result as the `AttributeUpdate`'s value.

We need to set the value here for all types of updates because we cannot serialize a function (to Tier1). Once the evaluation happens, the `AttributeUpdateByReference` essentially becomes a simple `AttributeUpdate` and act as before.

If you are worried about how this interacts with `Accumulate`, it will have no effect. We are just setting the same value twice on the object, with no side effects. The alternative would be to copy-paste these 3 lines to every other type of attribute update type in the above `switch` block, so I preferred to do it here.

The are unit tests that verify the by-ref with `Accumulate` combination, so this does work as intended.